### ISSUE TITLE
authorized_keys module fails when the public key has a comment

### ISSUE THREAD

**Issue by nueces** [ansible/ansible#5432]
Whit the new release 1.4.X, and the devel branch, if the public key has a comment the autorized_key module fails. This was previously working for as in the 1.3.X releases. 

Traceback:

```
TASK: [staging-common | Set user's authorized_keys] 
<staging> ESTABLISH CONNECTION FOR USER: testuser
<staging> EXEC ['ssh', '-tt', '-q', '-o', 'ControlMaster=auto', '-o', 
'ControlPersist=60s', '-o', 
'ControlPath=/home/nueces/.ansible/cp/ansible-ssh-%h-%p-%r', '-o', 
'Port=22', '-o', 'KbdInteractiveAuthentication=no', '-o', 
'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'User=testuser', '-o', 'ConnectTimeout=10', 'st-telesur-plone-edit-02', "/bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-1388089157.05-173159870032279 && chmod a+rx $HOME/.ansible/tmp/ansible-1388089157.05-173159870032279 && echo $HOME/.ansible/tmp/ansible-1388089157.05-173159870032279'"]
<staging> REMOTE_MODULE authorized_key user=nueces key="# nueces
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDrPq0e80i8332okNf4SD03NqOQ/hGWmsPC9COjJPEnWEO5NjvEQWaUbkPsbqP5FcqtBlLYRN3mEW13XwNx0gxX/UubkS2Raf0Vf8/+Jcf6i06IgQMZ1lerHu1Shr9fxZFNCF4l6qmmlVI891BWVhn3Vt3uLwM8DhB+tbdbdcoticOukp93BGUpXGKWifsdM4uPA0L+hxl4k8YZNlCMDyID4fUo+0raXY8Rs59ilt7pY7iTAnksTbbaoTCESo+b6DCuj5EVYSLQ742xxRnvH4Lx8w0vfOJXF5tRWPcQvspr//hc/y06B8LevjmnCzUAhQoFvYolPlJSEWQCIW3TlZDv nueces@sasha"
<staging> PUT /tmp/tmppc486M TO /home/testuser/.ansible/tmp/ansible-1388089157.05-173159870032279/authorized_key
<staging> EXEC ['ssh', '-tt', '-q', '-o', 'ControlMaster=auto', '-o', 
'ControlPersist=60s', '-o', 
'ControlPath=/home/nueces/.ansible/cp/ansible-ssh-%h-%p-%r', '-o', 
'Port=22', '-o', 'KbdInteractiveAuthentication=no', '-o', 
'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'User=testuser', '-o', 'ConnectTimeout=10', 'staging', '/bin/sh -c \'sudo -k && sudo -H -S -p "[sudo via ansible, key=ofojsqxugsxffcqrcebpljsnkoghtpwc] password: " -u root /bin/sh -c \'"\'"\'echo SUDO-SUCCESS-ofojsqxugsxffcqrcebpljsnkoghtpwc; /usr/bin/python /home/testuser/.ansible/tmp/ansible-1388089157.05-173159870032279/authorized_key; rm -rf /home/testuser/.ansible/tmp/ansible-1388089157.05-173159870032279/ >/dev/null 2>&1\'"\'"\'\'']
failed: [staging] => (item=nueces) => {"failed": true, "item": "nueces"}
msg: invalid key specified: # nueces
```

This was very usefull as inline documentation in the authorized_key file of the target user.

**Comment by mpdehaan** [comment]
Can you please show the line you have in your playbook and perhaps a sanitized version of the authorized key file?

**Comment by lichesser** [comment]
This can be reproduced with e.g. `- authorized_key: user=charlie key="{{ lookup('file', '/home/charlie/.ssh/id_rsa.pub') }}"`

where `id_rsa.pub` contains (took the data from the error report above)

```
# nueces
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDrPq0e80i8332okNf4SD03NqOQ/hGWmsPC9COjJPEnWEO5NjvEQWaUbkPsbqP5FcqtBlLYRN3mEW13XwNx0gxX/UubkS2Raf0Vf8/+Jcf6i06IgQMZ1lerHu1Shr9fxZFNCF4l6qmmlVI891BWVhn3Vt3uLwM8DhB+tbdbdcoticOukp93BGUpXGKWifsdM4uPA0L+hxl4k8YZNlCMDyID4fUo+0raXY8Rs59ilt7pY7iTAnksTbbaoTCESo+b6DCuj5EVYSLQ742xxRnvH4Lx8w0vfOJXF5tRWPcQvspr//hc/y06B8LevjmnCzUAhQoFvYolPlJSEWQCIW3TlZDv nueces@sasha
```

The content of the `authorized_key` file is irrelevant here.

It fails in `enforce_state` because every line of the `key` parameter is fed to `parsekey` [here](https://github.com/ansible/ansible/blob/872a1ee39b5651d8da3bdc2d7e22c64927add850/library/system/authorized_key#L336), which says the line

```
# nueces
```

is not a valid ssh key.

---

Some more information:

The `authorized_key` code currently supports the `key` parameter to be either one or more valid ssh keys seperated by `\n`.

I'm not entirely sure why the multi-key ability is even there (and it doesn't seem to be documented) as previously - see https://github.com/ansible/ansible/commit/39c8becfc87012f9241a6c0a83aaeb4bf73e73c5 - `authorized_key` even failed explicitly when `key` contained more then one line. So somewhere in between it supported multiple lines but didn't verify if they look like ssh keys. Which is the reason why line comments "worked" for @nueces until `parsekey` came along.

Line comments are a valid part of a key file (as are empty lines which probably also would cause a failure) while multiple keys in one key file are not supported as far as I know (`ssh-keygen -l -f keyfile` only recognizes the first key in a file).

I stop here noting that supporting line comments could be difficult as it's e.g. hard to know whether a comment changed or if this simply is a new comment. Also theoretically with the current code the placement off comment and ssh-key need not to be in the same place and order. e.g. exact same comment already there in other place of file or similar.

**Comment by sivel** [comment]
I believe this issue was resolved by #5655

**Comment by jctanner** [comment]
@sivel yep. @nueces please retest against the latest devel.
